src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,109 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_LENGTH = 10;
+    private int size;
+    private Object[] values;
+
+    public ArrayList() {
+        values = new Object[DEFAULT_LENGTH];
+    }
+
     @Override
     public void add(T value) {
-
+        checkLength();
+        values[size()] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size == index) {
+            add(value);
+            return;
+        }
+        size++;
+        checkValidIndex(index);
+        checkLength();
+        System.arraycopy(values, index, values, index + 1, size - index);
+        values[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkValidIndex(index);
+        return (T) values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkValidIndex(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkValidIndex(index);
+        T removedValue = (T) values[index];
+        System.arraycopy(values, index + 1, values, index, size - index - 1);
+        size--;
+        return removedValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getIndexValue(element);
+        T removedValue = (T) values[index];
+        System.arraycopy(values, index + 1, values, index, size - index);
+        size--;
+        return removedValue;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
     }
+
+    private void checkLength() {
+        if (values.length == size) {
+            getScaledArray();
+        }
+    }
+
+    private void getScaledArray() {
+        Object[] copyArray = new Object[values.length + (values.length >> 1)];
+        System.arraycopy(values, 0, copyArray, 0, size);
+        values = copyArray;
+    }
+
+    private void checkValidIndex(int index) {
+        if (index >= size() || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " is invalid");
```

Comment:
```suggestion
            throw new ArrayListIndexOutOfBoundsException("Index " + index + " is invalid for size " + size);
```

lets make exception messages more informative